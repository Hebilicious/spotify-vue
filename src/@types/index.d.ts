export type Track = {
    _id: string,
    name: string,
    year: number | null,
    artist: string | null,
    album: string | null,
    duration_s: number,
    imageUri: string,
    type: string,
    uri: string
};

export type Album = {
    type: string,
    _id: string,
    name: string,
    year: number,
    total_tracks: number,
    description: string,
    imageUri: string,
    tracks: Track[]
};

export type PlayList = Album[]