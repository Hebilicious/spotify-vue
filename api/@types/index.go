package api

type Track struct {
	ID       string `json:"_id"`
	Name     string `json:"name"`
	Year     int    `json:"year"`
	Artist   string `json:"artist"`
	Album    string `json:"album"`
	Duration int    `json:"duration_s"`
	ImageUri string `json:"imageUri"`
	Type     string `json:"type"`
	Uri      string `json:"uri"`
}

type Album struct {
	ID          string  `json:"_id"`
	Type        string  `json:"type"`
	Name        string  `json:"name"`
	Year        int     `json:"year"`
	TotalTracks int     `json:"total_tracks"`
	Description string  `json:"description"`
	ImageUri    string  `json:"imageUri"`
	Tracks      []Track `json:"tracks"`
}

type PlayList []Album
