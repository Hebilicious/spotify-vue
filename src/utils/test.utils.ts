import { VueWrapper } from '@vue/test-utils';
import { expect, it } from 'vitest';

export const testSnapShot = (wrapperGetter: () => VueWrapper) => {
  it('Matches the snapshot', () => {
    expect(wrapperGetter().html()).toMatchSnapshot();
  });
};
