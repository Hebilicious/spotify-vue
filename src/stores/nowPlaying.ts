import { defineStore } from 'pinia';

import { Track } from '@/@types';

export const useNowPlayingStore = defineStore({
  id: 'playback-queue',
  state: () => ({
    playQueue: ([] as Track[]),
    currentIndex: (undefined as unknown as number),
  }),
  getters: {
    getCurrentPlaylist: ({ playQueue }) => playQueue,
    getCurrentPlaying: ({ playQueue, currentIndex }) => currentIndex !== undefined ? playQueue?.[currentIndex] : (undefined as unknown as Track),
  },
  actions: {
    setCurrentPlaylist(newPlayQueue: Track[]) {
      this.$patch({ playQueue: newPlayQueue, currentIndex: 0 });
    },
    next() {
      (this.currentIndex < (this.playQueue.length - 1)) && this.currentIndex++;
    },
    prev() {
      this.currentIndex > 0 && this.currentIndex--;
    },
    clearPlaylist() {
      this.playQueue = [];
      this.currentIndex = (undefined as unknown as number);
    }
  },
});
