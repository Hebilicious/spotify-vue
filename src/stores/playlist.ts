import { PlayList } from '@/@types';
import { getAllPlayLists } from '@/services/playListService';
import { defineStore } from 'pinia';

export const usePlayListStore = defineStore({
    id: 'playlists',
    state: () => ({
        playlists: ([] as PlayList),
    }),
    getters: {
        getPlayLists: ({ playlists }) => playlists,
    },
    actions: {
        async loadPlayLists() {
            this.playlists = await getAllPlayLists();
        }
    },
});

export default usePlayListStore;