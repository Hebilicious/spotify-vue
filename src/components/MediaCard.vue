<template>
  <div class="media-card">
    <img :src="album.imageUri" alt="cover-art" />
    <div class="play-icon" @click="playCurrentAlbum">
      <div class="play-icon-circle">
        <span class="material-icons">play_arrow</span>
      </div>
    </div>
    <div class="media-title">{{ album.name }}</div>
    <span class="media-subtitle">{{ album.description }}</span>
  </div>
</template>

<script lang="ts" setup>
import { Album } from '@/@types';
import { useNowPlayingStore } from '@/stores/nowPlaying';

const nowPlayingStore = useNowPlayingStore();
const { album } = defineProps<{ album: Album }>();

const playCurrentAlbum = () => {
  nowPlayingStore.setCurrentPlaylist(album.tracks)
}
</script>

<style lang="scss">
.media-card {
  background-color: var(--color-background-soft);
  height: 250px;
  width: 185px;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  padding: 1rem;
  align-items: flex-start;
  transition: all 0.5s ease;

  img {
    height: 150px;
    width: 100%;
    border-radius: 5px;
    object-fit: cover;
  }

  .media-title {
    padding: 1rem 0.5rem 0.25rem 0;
    white-space: nowrap;
    width: 100%;
    overflow: hidden;
    text-overflow: ellipsis;
    font-size: 16px;
    font-weight: 700;
    letter-spacing: normal;
    line-height: 24px;
    text-transform: none;
    color: var(--color-heading);
  }

  .play-icon {
    visibility: hidden;
    transform: translate(110px, 110px);
    position: absolute;
    opacity: 0;
    cursor: pointer;
    transition: visibility 0.5s ease, opacity 0.5s ease-in-out;

    .play-icon-circle {
      border-radius: 50%;
      background: #1fdf64;
      height: 50px;
      width: 50px;
      box-shadow: 0 8px 8px rgba(0, 0, 0, 0.3);

      transform: translate(-10px, -15px);
      span {
        transition: transform 0.5s ease-out;
        color: black;
        transform: translate(13px, 13px) scale(1);
      }
    }
  }

  &:hover {
    background-color: #252424;
    transform: scale(1.05);
    .play-icon {
      visibility: visible;
      opacity: 1;
      span {
        transform: translate(13px, 13px) scale(2);
      }
    }
  }
}
</style>
