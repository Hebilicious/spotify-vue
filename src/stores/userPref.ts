import { Track } from '@/@types';
import { defineStore } from 'pinia';
import { useNowPlayingStore } from './nowPlaying'

export const useUserPrefStore = defineStore({
    id: 'user-preferences',
    state: () => ({
        likedSongs: ([] as Track[]),
        volumePref: 0.6
    }),
    getters: {
        getLikedSongs: ({ likedSongs }) => likedSongs,
        isLikedSong: ({ likedSongs }) => {
            const playbackStore = useNowPlayingStore();
            return playbackStore.getCurrentPlaying && likedSongs.find(({ _id }) => _id === playbackStore.getCurrentPlaying._id)
        },
        getVolumePref: ({ volumePref }) => volumePref
    },
    actions: {
        toggleLikedSong(track: Track) {
            this.likedSongs = this.isLikedSong
                ? this.likedSongs.filter(({ _id }) => _id !== track._id)
                : [...this.likedSongs, track]
        },

        setVolumePref(volumePref: number) {
            this.volumePref = volumePref;
        }
    },
});
