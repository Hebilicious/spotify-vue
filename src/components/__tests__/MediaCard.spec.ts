import { beforeEach, describe, expect, fn, it } from 'vitest';
import { shallowMount, VueWrapper, } from '@vue/test-utils';
import { createTestingPinia } from '@pinia/testing'

import MediaCard from '@/components/MediaCard.vue';

import { testSnapShot } from '@/utils/test.utils';
import { useNowPlayingStore } from '@/stores/nowPlaying';
import { Album } from '@/@types';

describe('MediaCard', () => {
  let wrapper: VueWrapper;
  const mockMediaCardProps: Partial<Album> = {
    imageUri: 'mock-url',
    name: 'mock-title',
    description: 'mock-subtitle',
    tracks: []
  };

  beforeEach(() => {
    wrapper = shallowMount(MediaCard, {
      props: { album: (mockMediaCardProps as Album) },
      global: {
        plugins: [
          createTestingPinia({
            createSpy: () => fn()
          })
        ]
      }
    });
  });

  it('displays the details passed', () => {
    expect(wrapper.find('.media-card img').attributes('src')).toBe('mock-url');
    expect(wrapper.find('.media-title').text()).toBe('mock-title');
    expect(wrapper.find('.media-subtitle').text()).toBe('mock-subtitle');
  });

  it('calls the loadPlaylist from store when play button is clicked', async () => {
    const store = useNowPlayingStore();
    await wrapper.find('.play-icon').trigger('click');

    expect(store.setCurrentPlaylist).toBeCalledWith(mockMediaCardProps.tracks);
  })

  testSnapShot(() => wrapper);
});
