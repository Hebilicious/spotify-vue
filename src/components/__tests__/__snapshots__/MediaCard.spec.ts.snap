// Vitest Snapshot v1

exports[`MediaCard > Matches the snapshot 1`] = `
"<div class=\\"media-card\\"><img src=\\"mock-url\\" alt=\\"cover-art\\">
  <div class=\\"play-icon\\">
    <div class=\\"play-icon-circle\\"><span class=\\"material-icons\\">play_arrow</span></div>
  </div>
  <div class=\\"media-title\\">mock-title</div><span class=\\"media-subtitle\\">mock-subtitle</span>
</div>"
`;
